Add a PIDFile class that can be used as a context

glyph opened the pull request:
https://twistedmatrix.com/trac/ticket/8659

wsanchez committed:
docstring

wsanchez committed:
__all__ isn't appropriate, as nothing is exported here.

wsanchez committed:
Clean up DummyFilePath:
 * Default content is no content
 * Make sure self._exists and self._content are consistent

wsanchez committed:
Add PIDFile class and tests

wsanchez committed:
Remove some cruft

wsanchez committed:
Add context manager

wsanchez committed:
Shouldn't be allowed to kill PID 1, but it should still register as running

wsanchez committed:
Add PIDFile.format()

wsanchez committed:
Test for specific expected errno cases: ESRCH, EPERM.  Also EEXIST as an unexpected case.

wsanchez committed:
Add test_isRunningDoesExist, which patches kill.
This way we are testing both assuming kill works as we expect, and kill however it actually works on the running platform.

wsanchez committed:
Add test_isRunningInit, which tests for a running init process on POSIX systems.

wsanchez committed:
Add tests for PIDFile.format()

wsanchez committed:
Use PIDFile

wsanchez committed:
Move DummyFilePath

wsanchez committed:
doc cleanup

wsanchez committed:
isRunning should return false, not raise, if there is no PID file

wsanchez committed:
Raise an exception in context manager if PID is already running

wsanchez committed:
Add nonePIDFile

wsanchez committed:
PIDFile.read() returns None if the file doesn't exist instead of raising.

wsanchez committed:
Exit if already running.

wsanchez committed:
Make sure PID file is used

wsanchez committed:
doc fix

wsanchez committed:
Tests had swapped implementations

markrwilliams commented on src/twisted/application/runner/_pidfile.py:
```diff
@@ -0,0 +1,202 @@
+        """
+        pidString = b""
+        try:
+            for pidString in self.filePath.open():
```
This will retain a file descriptor for an indeterminate amount of time under PyPy.  Best to use a context manager!

markrwilliams commented on src/twisted/application/runner/_pidfile.py:
```diff
@@ -0,0 +1,202 @@
+        @raise EnvironmentError: If this PID file cannot be read.
+        @raise ValueError: If this PID file's content is invalid.
+        """
+        pidString = b""
```
Is this the empty string so that an empty file raises the same error as a file with a non-integral PID?  Would it be better to raise a more informative error message in the case of an empty file?  I think I'd appreciate the extra help debugging such a weird occurrence.

glyph committed:
Merge branch 'trunk' into 8659-pidfile

glyph commented:
Needs a topfile so the builders will be happy…

glyph commented:
Also, looks like the Windows build is failing?

glyph committed:
Merge branch 'trunk' into 8659-pidfile

glyph committed:
Merge branch 'trunk' into 8659-pidfile

moshez commented on src/twisted/application/runner/_pidfile.py:
```diff
@@ -0,0 +1,202 @@
+
+
+    @staticmethod
+    def format(pid=None):
```
Why does this need to allow None in?

wsanchez replied:
Yes… that was because… `write()` takes a `None` pid and probably at one point that made sense but there's no longer a need for that here.  Removing.

wsanchez replied:
6098f0c83bb1f1b8153fe04f6ed2ab1f8c7c882d

moshez commented on src/twisted/application/runner/_pidfile.py:
```diff
@@ -0,0 +1,202 @@
+        @param filePath: The path to the PID file on disk.
+        @type filePath: L{IFilePath}
+        """
+        self.filePath = IFilePath(filePath)
```
If filePath is already supposed to implement IFilePath, why are we doing adaptation here?

wsanchez replied:
I guess I don't know the right way to right "object adaptable to IFilePath" in the docstring without sounding too verbose, or what the best practice for that is.

glyph replied:
The evolving convention here, I think, is … don't do adaptation "inside" an API like this.  Require that the caller pass an `IFilePath` themselves, and if they want to do adaptation on their end, they can do it.

Adaptation is convenient, but it introduces coupling to the global state of the adapter registry, which in turn potentially has implications for threading, and generally we don't have great tools for telling us why an adaptation did or didn't work.  Generally the places where I find adaptation still makes sense are big heterogenous datastructures where a "walker" does adaptation over fragments of it (like twisted.web.template); if it's just an individual parameter, I almost never do adaptation directly like this any more.

wsanchez replied:
498f517795d5635152447f0bc768a94c97124625 removes adaptation

moshez commented on src/twisted/application/runner/_pidfile.py:
```diff
@@ -0,0 +1,202 @@
+        return int(pidString)
+
+
+    def write(self, pid=None):
```
Can this be separated into

```
def writeMyPID(self):
```

and

```
def write(self, pid):
       # Does not allow None here
```

wsanchez replied:
062493182c003280c20b016ca343bf6ee9a3028c - Adds writeRunningPID()

moshez commented on src/twisted/application/runner/_pidfile.py:
```diff
@@ -0,0 +1,202 @@
+        """
+        Read the process ID stored in this PID file.
+
+        @return: The contained process ID, or C{None} if the PID file does not
```
This can be a dangerous API -- users must explicitly test for None, and are likely to not do this in test scenarios. Can we have this API instead raise a custom (or standard) exception like ValueError to indicate "no PID"?

wsanchez replied:
Agreed.  I think I like custom, so we're very explicit.
- 5619da41be4a5f21a67c76dcdac25409e8561446  - Adds InvalidPIDFileError custom exception for the cases where the PID file has invalid contents.
- 71781a7991f38faca4d26e6564ecc6c73ebd0067 - Adds NoPIDFound custom exception for when there is no PID (non-existing PID file).

moshez commented on src/twisted/application/runner/_pidfile.py:
```diff
@@ -0,0 +1,202 @@
+        except OSError as e:
+            if e.errno == errno.ESRCH:  # No such process
+                self._log.info("Removing stale PID file: {log_source}")
+                self.remove()
```
Potentially weird side-effect: a boolean function modifying global (filesystem) state. I realize an invalid PID is equivalent to an empty file in some respect, but this feels like it can cause hard-to-debug problems.

wsanchez replied:
I was conflicted about this myself; I prefer for changed to the underlying file to be explicit (ie. limited to `write()`, `remove()` and the like).

The reason I did this here anyway is that the PID file does not reference a valid PID right now, and there's a risk that another PID may spin up with that PID later, which is potentially an even nastier problem, so I felt like being proactive about it.

Another option here is to raise an exception and force the caller to deal with this.  I didn't go for that because I suspect that this would mean that all callers should basically add more code to catch this exception and… remove the file; otherwise, this case would raise an uncaught exception.

Thoughts?

wsanchez replied:
OK so:

4351561fb484d3f5919711e29336361128f1406a - Adds StalePIDFileError and raises that if the PID file is stale.  Caller handles it.  If we need to re-encapsulte that, options include a wrapper function, or a kwarg asking for this explicitly.

moshez commented on src/twisted/application/runner/_pidfile.py:
```diff
@@ -0,0 +1,202 @@
+    when no PID file is desired.
+    """
+
+    def __init__(self):
```
This is redundant.

wsanchez replied:
No, because superclass takes an argument…

markrwilliams replied:
Doesn't this no-argument initializer mean that `NonePIDFile` doesn't implement the `IPIDFile` interface?  If so, is this OK because we prefer to verify that an object provides an interface to verifying that its class implements it?

If this should take an argument to ensure that the class implements `NonePIDFile`, is a default value of `None` acceptable and a note in the doc string that the argument is ignored?

moshez commented on src/twisted/application/runner/_pidfile.py:
```diff
@@ -0,0 +1,202 @@
+
+
+
+class NonePIDFile(PIDFile):
```
Why the inheritance? Should we have an explicit interface for a PIDFile?

wsanchez replied:
I did that for isinstance to work, but an interface would be more cool.

wsanchez replied:
IPIDFile added in d36c4695c6fdd246087afad3351f9577c8ae670f, inheritance removed.

moshez commented on src/twisted/application/runner/_pidfile.py:
```diff
@@ -0,0 +1,202 @@
+
+
+
+nonePIDFile = NonePIDFile()
```
Should this class be private, or do we expect allowing people to construct new nonefiles?

wsanchez replied:
Everything here is currently private…

moshez commented on src/twisted/application/runner/_runner.py:
```diff
@@ -46,48 +42,41 @@ def __init__(self, options):
         if self.options.get(RunnerOptions.kill, False):
-            if pidFilePath is None:
-                exit(ExitStatus.EX_USAGE, "No PID file specified")
+            if pidFile is None:
```
This will never be None, right? It will be the NonePIDFile?

wsanchez replied:
Presently it could be None, but the next ticket will rework options handling, so let me punt on this issue until then.

moshez commented on src/twisted/application/runner/_runner.py:
```diff
@@ -201,9 +169,9 @@ class RunnerOptions(Names):
-    @cvar pidFilePath: The path to the PID file.
-        Corresponding value: L{IFilePath}.
-    @type pidFilePath: L{NamedConstant}
+    @cvar pidFile: The path to the PID file.
```
It's not "the path" anymore, right?

wsanchez replied:
Correct.  8a772253d32ec2db6f19f211fbbb26eeca061229

moshez commented on src/twisted/application/runner/test/test_pidfile.py:
```diff
@@ -0,0 +1,291 @@
+        def kill(pid, signal):
+            return  # Don't actually kill anything
+
+        self.patch(_pidfile, "kill", kill)
```
I don't know how I feel about this, but I'm not sure I have a better alternative.

wsanchez replied:
I feel OK about this.  :-)

moshez commented on src/twisted/application/runner/test/test_pidfile.py:
```diff
@@ -0,0 +1,291 @@
+            raise SkipTest("This test assumes POSIX")
+
+        pidFile = PIDFile(DummyFilePath())
+        pidFile.write(1)  # PID 1 is init on POSIX systems
```
Running the tests in a typical docker container would not expose a problem with EPERM. I suggest using the same patching strategy used above.

wsanchez replied:
That would make this identical to `test_isRunningNotAllowed`.

markrwilliams replied:
More generally this will fail if the tests run as `root` (which is, of course, generally nuts.)  You could add a skip for this test that checks if it's running on a POSIX system and the process' effective user ID is 0 to avoid spurious failures in Docker containers.

moshez commented on src/twisted/application/runner/test/test_pidfile.py:
```diff
@@ -0,0 +1,291 @@
+
+
+
+class DummyFilePath(FilePath):
```
My usual attitude is that using real files is not a big problem since filesystems are usually reliable and you can have files that are local to the test environment and more-or-less guaranteeed-uniguely name. Not sure if this stub is worth the extra complexity?

moshez commented on src/twisted/application/runner/test/test_pidfile.py:
```diff
@@ -0,0 +1,291 @@
+
+
+    def open(self, mode="r"):
+        if not self._exits:
```
Is this supposed to be "exists"? (Ditto for below)

wsanchez replied:
Yes: c19e9d847afd2a8337bccd0e3cdf7b5c2d846e74

moshez commented on src/twisted/application/runner/test/test_runner.py:
```diff
@@ -367,6 +375,58 @@ def exited(options):
 
 
 
+class DummyRunner(Runner):
```
Do we need inheritance here?

wsanchez replied:
No, it's just expedient.

moshez commented on src/twisted/application/runner/test/test_runner.py:
```diff
@@ -367,6 +375,58 @@ def exited(options):
+
+
+
+class DummyPIDFile(NonePIDFile):
```
Ditto -- do we need inheritance?

wsanchez replied:
No, it's just expedient.

moshez reviewed (approved):
Thanks a lot for doing this -- I am a big fan of this refactor. This is much clearer than it used to be.

I made a bunch of comments, but I don't think any single one is a show-stopper. Please address those you think that have merit and either merge directly or put up for review if you think the changes are significant.

wsanchez committed:
No need for format() to take None

wsanchez committed:
Link EnvironmentError and ValueError.

wsanchez committed:
Add InvalidPIDFileError custom exception.

wsanchez committed:
Add NoPIDFound custom exception for valid cases where there is no PID (there's no PID file).
Add test for non-existing PID file case.

wsanchez committed:
Add writeRunningPID()

wsanchez committed:
Reverse c27e346b9c68db308b949a73d661a155e4435159, whoops.

wsanchez committed:
Raise StalePIDFileError if isRunning() encounters a stale PID file.

wsanchez committed:
Don't adapt to IFilePath

wsanchez committed:
Doc fix for pidFile option.

wsanchez committed:
typo

wsanchez committed:
Merge branch 'trunk' into 8659-pidfile

wsanchez committed:
Merge branch '8659-pidfile' of github.com:twisted/twisted into 8659-pidfile

wsanchez committed:
Add topfile

wsanchez committed:
Merge branch 'trunk' into 8659-pidfile

wsanchez committed:
Merge branch 'trunk' into 8659-pidfile

wsanchez committed:
Add test for read re-raising OSError if the associated errno is not ENOENT.

wsanchez committed:
Fix docstring; we no longer return None.

wsanchez committed:
Match NonPIDFile.read() to PIDFile.read(); don't return None.

wsanchez committed:
Also check errno

wsanchez committed:
Add tests for NonePIDFile.

wsanchez committed:
pid arg to write() should not be optional.

wsanchez committed:
Add IPIDFile interface.
PIDFile and NonePIDFile implement IPIDFile.
NonePIDFile no longer inherrits from PIDFile.

wsanchez committed:
Don't catch BrokenMethodImplementation because coverage

wsanchez committed:
Add test for NonePIDFile.writeRunningPID

wsanchez committed:
Call self.getContent() instead of reaching for self._content.

wsanchez committed:
Call __enter__ explicitly because coverage.

wsanchez committed:
Merge branch 'trunk' into 8659-pidfile

glyph committed:
Merge branch 'trunk' into 8659-pidfile

markrwilliams commented on src/twisted/application/runner/_pidfile.py:
```diff
@@ -0,0 +1,267 @@
+    def read(self):
+        pidString = b""
+        try:
+            for pidString in self.filePath.open():
```
This will leak a file descriptor/handle under PyPy and produce a `ResourceWarning` under Python 3.  Wrapping the `open` call with context manager would be best though it'll make the indentation even deeper.

wsanchez replied:
Fixed: bdd6fac18c7c6abc150bcc6ffcc1cdd4eff3b068

markrwilliams commented on src/twisted/application/runner/_pidfile.py:
```diff
@@ -0,0 +1,267 @@
+        try:
+            return int(pidString)
+        except ValueError:
+            raise InvalidPIDFileError("non-integer PID value in PID file")
```
It'd be nice to see the repr of the invalid value in the exception message.

wsanchez replied:
Fixed: a486e26bb287daef819be3f864fbe7c27d502b19
Also added a test, and noticed one of the read tests wasn't right, so fixed that as well.

markrwilliams commented on src/twisted/application/runner/test/test_pidfile.py:
```diff
@@ -0,0 +1,416 @@
+        L{NonePIDFile} conforms to L{IPIDFile}.
+        """
+        pidFile = NonePIDFile()
+        verifyObject(IPIDFile, pidFile)
```
I guess this answers the implements vs provides verificatio :)

wsanchez replied:
Yeah, the interface doesn't define the initializer; just the behavior of the instantiated object.

wsanchez replied:
geteuid test skippage added in 6fe8c9d28a10480191461b8e622db5cd7e2316b8

markrwilliams reviewed (commented):
Resource leak, more explanation in a exception message, some interface questions, and a suggestion for testing in Docker containers.

wsanchez committed:
Merge branch 'trunk' into 8659-pidfile

wsanchez committed:
Use a context manager so we don't leak a fine handle.

wsanchez committed:
Add repr of invalid PID file content when raising InvalidPIDFileError.

wsanchez committed:
Check exception strings.

wsanchez committed:
Skip test_isRunningInit if process euid is 0.

wsanchez committed:
str->unicode

wsanchez committed:
Try again with 3.5 string handling.  Problem was repr or str, not str vs. unicode.

wsanchez committed:
Don't import geteuid if not POSIX

wsanchez committed:
Merge branch 'trunk' into 8659-pidfile

wsanchez commented:
Reverted…
